packages/framework/src/Component/Elasticsearch/IndexRepository.php

```diff
@@ -0,0 +1,329 @@
+    }
+
+    /**
+     * @param string $indexAlias
+     * @param array $data
+     * @param bool $createIfNotExists
+     */
+    public function bulkUpdate(string $indexAlias, array $data, bool $createIfNotExists = true): void
+    {
+        $params = ['body' => []];
+
+        foreach ($data as $id => $row) {
+            $params['body'][] = [
+                'update' => [
+                    '_index' => $indexAlias,
+                    '_type' => '_doc',
+                    '_id' => (string)$id,
+                ],
+            ];
+
+            $params['body'][] = [
+                'doc' => $row,
+                'doc_as_upsert' => $createIfNotExists,
+            ];
+        }
+
+        $result = $this->elasticsearchClient->bulk($params);
+
+        if (isset($result['errors']) && $result['errors'] === true) {
+            throw ElasticsearchIndexException::bulkUpdateError($indexAlias, $result['items']);
+        }
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Elasticsearch\AbstractIndex $index
+     * @param \Shopsys\FrameworkBundle\Component\Elasticsearch\IndexDefinition $indexDefinition
+     * @param \Symfony\Component\Console\Output\OutputInterface $output
+     */
+    public function exportAll(AbstractIndex $index, IndexDefinition $indexDefinition, OutputInterface $output): void
+    {
+        $this->sqlLoggerFacade->temporarilyDisableLogging();
+
+        $indexAlias = $indexDefinition->getIndexAlias();
+        $domainId = $indexDefinition->getDomainId();
+        $progressBar = $this->progressBarFactory->create($output, $index->getTotalCount($indexDefinition->getDomainId()));
+
+        $exportedIds = [];
+        $lastProcessedId = 0;
+        do {
+            // detach objects from manager to prevent memory leaks
+            $this->entityManager->clear();
+            $currentBatchData = $index->getExportDataForBatch($domainId, $lastProcessedId, $index->getExportBatchSize());
+            $currentBatchSize = count($currentBatchData);
+
+            if ($currentBatchSize === 0) {
+                break;
+            }
+
+            $this->bulkUpdate($indexAlias, $currentBatchData);
+            $progressBar->advance($currentBatchSize);
+
+            $exportedIds = array_merge($exportedIds, array_keys($currentBatchData));
+            $lastProcessedId = array_key_last($currentBatchData);
+        } while ($currentBatchSize >= $index->getExportBatchSize());
+
+        $this->deleteNotPresent($indexDefinition, $exportedIds);
+
+        $progressBar->finish();
+        $output->writeln('');
+
+        $this->sqlLoggerFacade->reenableLogging();
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Elasticsearch\AbstractIndex $index
+     * @param \Shopsys\FrameworkBundle\Component\Elasticsearch\IndexDefinition $indexDefinition
+     * @param int[] $restrictToIds
+     */
+    public function exportIds(AbstractIndex $index, IndexDefinition $indexDefinition, array $restrictToIds): void
```

Comment:
I tested the following scenario
- all products are exported in ES
- I go to product detail and remove all category assignments on the first domain and save product

Expected result: The product is no longer visible in product lists on the first domain.
Actual result: Product is visible in list and product detail returns 404

In the previous implementation was some condition to remove ids from elastic if they're not found with detExportData method.

Reply:
Fixed in https://github.com/shopsys/shopsys/pull/1622/commits/d9ad0cdeb3f3c663cdc93205baa312ecf6dbb3b9. Really thanks for noticing 👍